### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by Keats** [Keats/tera2#4]
The big-table especially is slower than I'd expect.

**Comment by Keats** [comment]
I'm not entirely sure how to speed it up. I've tried using an iterator, getting by idx everytime and it doesn't move the needle.

**Comment by jalil-salame** [comment]
On `big-table`, tera seems to spend ~30-40% of the time in `fmt::write`, so I don't think there is much to be done optimizing it further in isolation, maybe trying to reduce teh size of `Instruction` could pay off?

For example, if a `String` or a `Vec` will not grow in size, consider using a `Box<str>` or `Box<[T]>` instead (2*word size instead of 3).

Maybe also providing a bigger benchmark (a real world template which takes a long time to execute).

**Comment by Keats** [comment]
I was thinking of using `str` in `Value` as well if possible.
To be honest I think the big-team template is kind of the worst case template. I will try to add some more realistic benchmark (eg with inheritance, macros, loops, set etc)

**Comment by Keats** [comment]
More realistic set of templates: https://github.com/Keats/tera2/pull/11

**Comment by onkoe** [comment]
@Keats, I see! I'm going to post my whole problem, then. Please don't forget to mark it as unrelated if you find it unhelpful. 

Here's the 'real version' of the example I gave:

```html
{% macro button(label, type="default", color="primary", href="#", other_classes="", left-image="", right-image="") %} 

<link rel="stylesheet" href="{{ get_url(path="button.css", trailing_slash=false) }}"/>

{% if type == "default" %}
    <div class="btn-wrapper btn-wrapper-{{color}} {{ other_classes }}">
        <a class="btn" href="{{ href }}">
            <div>{{label}}</div>
        </a>
    </div>
{% elif type == "extended-fab" %}
    <div class="btn-wrapper btn-wrapper-{{color}} {{ other_classes }}">
        <a class="extended-fab-btn" href="{{ href }}">
            {% if left_image != "" %}
                <img src="{{get_url(path=left_image, trailing_slash=false) | safe}}" />
            {% endif %}

            <div>{{label}}</div>

            {% if right_image != "" %}
                <img src="{{get_url(path=right_image, trailing_slash=false) | safe}}" />
            {% endif %}

        </a>
    </div>
{% endif %}

{% endmacro button %}
```

After doing this, I get a funky little error message from Zola, requesting that my default values be boolean only. 

```console
Change detected @ 2023-10-19 19:09:43
-> Template changed /Users/barrett/Documents/barretts-club/templates/macros/button.html
Reloading only template
Error: Failed to build the site
Error:
* Failed to parse "/Users/barrett/Documents/barretts-club/templates/macros/button.html"
 --> 1:81
  |
1 | {% macro button(label, type="default", color="primary", href="#", other_classes="", left-image="", right-image="") %}
  |                                                                                 ^---
  |
  = expected `true` or `false`
Done in 7ms.
```

If I abide by its suggestions and use a boolean instead, it proceeds to go after my other optional arguments, such as `href="#"`.  

Is this more of a Zola issue, or am I misusing the macro system? Thanks for your quick response, by the way! 🤩

**Comment by Keats** [comment]
Another thing: we should probably keep Arc<str> in `Key` since otherwise we clone the string everytime we look up an attribute i believe...?

**Comment by Keats** [comment]
Another way to speedup things is to do what [ramhorns](https://maciej.codes/2019-03-03-ramhorns.html) is doing: pre-compute hashes for all struct fields. Ramhorns uses a derive but we could pre-compute the hash when compiling the templates and serializing to Value to avoid the derive. We would need to differentiate between dynamic and static maps/structs though but in practice, i'm guessing most people pass struct to the Context so we should be able to skip most hashmap lookups (well, we can just have a NoHashHasher or something for the HashMap with pre-computed hash).

I don't know when i'll have the time to try that as it can be a potentially big change. What do you think @jalil-salame ?

**Comment by jalil-salame** [comment]
I need to take a look at the profile and see how much hashing is taking up... My guess is we could make a Cow type but for hashing: either precomputed or lazily computed.

Hashing did not take a significant amount of time, but it could unlock optimizations by reducing the amount of code in the precomputed hash path.

**Comment by Keats** [comment]
Maybe there are some other optimizations ramhorns is doing but it does beat askama which is compiled to Rust so it's doing things correctly!

**Comment by Keats** [comment]
@jalil-salame did you end up having a look at it?

**Comment by jalil-salame** [comment]
> @jalil-salame did you end up having a look at it?

Not really, I have been busy :c, I don't think I have time to look at it for the next month or two either...

**Comment by Keats** [comment]
No worries I barely have time myself
